### PR TITLE
Add chair console voice dictation: mic button + Web Speech API

### PR DESCRIPTION
**Pull request by kdr**
## What

Voice control v1 for overcast: speak steer/follow-up prompts from the **/chair phone console**. A `mic` button in the composer streams live browser speech recognition (interim + final) into the textarea for review-then-send.

**Console-only change** — transcripts ride the existing `POST /api/prompt {text, mode}` path, so the bridge, wire contract, token auth, and steer/followUp/abort semantics are all untouched.

## How

- **`web/chair/src/dictation.ts`** (new): thin Web Speech API wrapper.
  - Fresh recognizer per session (reuse after `stop()` is flaky across engines), `continuous` + `interimResults`, `lang` from the browser locale.
  - Rebuilds finals+interim from the **full** result list every event — idempotent, and iOS Safari replaces earlier entries rather than only appending.
  - Graceful `stop()` (final flush still applies) vs hard `cancel()` (mutes late events) — submit uses `cancel()` so a late final flush can't repopulate the cleared box.
  - Support tiers: `ok` / `insecure` (API present, no secure context) / `unsupported` (no API — e.g. Firefox default).
- **`composer.ts`**: tap-to-toggle mic; while listening the textarea shows `typed base + finals + interim` live with the existing autosize; manual stop keeps the text for editing; button hidden when unsupported, blocked-with-hint when insecure. New optional `onNotice` handler surfaces dictation errors as transcript notices.
- **`theme.css`**: mic states in the csi palette (pulsing `● ` while listening, muted when blocked).
- **Docs**: README chair section, flows.md flow 23 (browser support, on-device vs cloud engines, the secure-context requirement and `tailscale serve` + Origin↔Host proxy note), CLAUDE.md `/chair` line.

## Why browser-side speech

- The chair console already models everything voice needs (auto/steer/followUp lanes, ABORT, token auth) — this is the smallest seam.
- The browser owns the mic permission prompt, sidestepping the macOS TCC footgun where terminal-launched processes silently record nothing.
- Browsers only grant the mic on secure origins: works on `localhost` as-is; a plain-HTTP tailnet bind shows the blocked hint (fix: HTTPS in front, e.g. `tailscale serve`).

## Verification

- `npm run typecheck`, `npm run build:web`, all **943 unit tests** pass, doc-references test passes. No server change → chair bridge tests untouched.
- **16-check Playwright end-to-end**: real `ChairBridge` (fake agent) serving the actual built console in headless Chromium with a mocked `SpeechRecognition`: support detection, listening state, interim→final transcript flow appending to typed text, stop-keeps-text, dictated prompt arriving at the agent through the real HTTP path, mid-dictation submit cancel + late-flush muting, button hidden when the API is missing. All pass.

## Out of scope (researched, deferred)

Desk-side push-to-talk in the TUI (local whisper.cpp / Groq batch / Apple Speech), always-on streaming STT (Deepgram / sherpa-onnx) + wake word, speech-to-speech concierge over the chair API, and a MediaRecorder→server-STT fallback route for browsers without the Web Speech API.

---

> [!NOTE]
> **Low Risk**
> Chair bridge auth and prompt paths are unchanged; additions are client-side speech, optional Tailscale CLI orchestration, and QR URL selection with guarded serve teardown.
> 
> **Overview**
> **Voice on the phone console** — a `mic` button in the chair composer streams Web Speech API transcripts into the textarea (review, then send/steer on the existing `POST /api/prompt` path). New `web/chair/src/dictation.ts` handles secure-context vs insecure vs unsupported; plain HTTP shows `mic 🔒` with guidance. `composer.ts` wires tap-to-toggle dictation, cancels on submit/abort/auth teardown, and surfaces errors via `onNotice`.
> 
> **HTTPS for pairing** — browsers need HTTPS (or localhost) for the mic. `ChairBridge` gains optional `publicUrl`, `displayUrl`, `secure`, and pairing URLs use `displayUrl` instead of raw `http://` bind. New `src/chair/serve.ts` integrates `tailscale serve`: detect existing HTTPS proxy to the chair port, enable with `--serve`, and tear down only mappings overcast owns. `/chair` adds `--serve`, `--url`, and `OVERCAST_CHAIR_URL` / `OVERCAST_TAILSCALE_CMD`; failed `--serve` does not disturb a running tailnet chair; port rebinds clean up orphaned serve mappings.
> 
> Docs (`README`, `flows.md`, `CLAUDE.md`) and unit tests cover serve parsing, bridge URLs, and extension edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8528a32d2061619ed62b8be09a673c8abffadaca. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: web/chair/src/dictation.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    };
+    rec.onend = () => {
+      // fires after stop(), abort(), errors, AND engine silence timeouts
+      if (!live) return;
+      live = false;
+      handlers.onState(false);
+    };
+    live = true;
+    handlers.onState(true);
+    rec.start();
```
[the remark this answers is not in the document]
Comment (kdr) [reply web/chair/src/dictation.ts]: Fixed in d04cc3f. `start()` set `live`/`onState(true)` before `rec.start()`, so a synchronous engine throw (e.g. `InvalidStateError`) left listening stuck with no `onend`. It now wraps `rec.start()` in try/catch and rolls the optimistic state back (`live=false` + `onState(false)` + an error notice) — matching the existing guard around `stop()`/`abort()` in `end()`. Covered by a new e2e check: a throwing `start()` leaves the mic off and surfaces a notice.

File: web/chair/src/views/composer.ts [hunk 2]
```diff
@@ -57,16 +105,14 @@ export function createComposer(handlers: {
       void submit();
     }
   });
-  input.addEventListener("input", () => {
-    input.style.height = "auto";
-    input.style.height = `${Math.min(input.scrollHeight, innerHeight * 0.3)}px`;
-  });
+  input.addEventListener("input", autosize);
```
[the remark this answers is not in the document]
Comment (kdr) [reply web/chair/src/views/composer.ts]: Fixed in d04cc3f. The ABORT handler now calls `dictation.cancel()` before `onAbort()`, mirroring `submit()` — so both terminal composer actions stop the mic. Covered by new e2e checks: ABORT turns the mic off, hard-cancels the recognizer, and still reaches the agent.

File: web/chair/src/dictation.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    rec.onerror = (e) => {
+      const message = ERROR_TEXT[e.error];
+      if (live && message !== null) handlers.onError(message ?? `dictation failed: ${e.error}`);
+    };
+    rec.onend = () => {
+      // fires after stop(), abort(), errors, AND engine silence timeouts
+      if (!live) return;
+      live = false;
+      handlers.onState(false);
+    };
```
[the remark this answers is not in the document]
Comment (kdr) [reply web/chair/src/dictation.ts:121]: Fixed in f616f51. Root-caused this together with the toggle-blocked report below — both came from the shared `live` flag being mutated by async callbacks from a stale recognizer. Reworked the state model: `listening()` now derives from the current recognizer ref (`rec`), and every callback is scoped to the instance that registered it (`rec === self`), so a superseded session's late `onend`/`onresult`/`onerror` are no-ops. This mirrors the `resyncToken` generation guard `main.ts` already uses for the SSE resync race. New e2e check: after stop→start, a delayed `onend` from the replaced recognizer no longer kills the new session (and the engine's own `onend` still ends the current one).

File: web/chair/src/views/composer.ts [hunk 4]
```diff
@@ -38,7 +42,51 @@ export function createComposer(handlers: {
...
+        "mic needs HTTPS (or localhost) — expose the chair with e.g. `tailscale serve` instead of the plain-HTTP tailnet bind",
+        "warning",
+      );
+      return;
+    }
+    if (dictation.listening()) {
+      dictation.stop();
+    } else {
+      base = input.value.trim();
+      dictation.start();
```
[the remark this answers is not in the document]
Comment (kdr) [reply web/chair/src/views/composer.ts:92]: Fixed in f616f51 as part of the same state-model rework (see the stale-onend thread). `stop()`/`cancel()` now flip listening OFF synchronously by clearing the current recognizer ref, so `listening()` is accurate immediately and the composer toggle starts a fresh session on the next tap instead of re-stopping; the just-ended instance's late events are dropped by the identity guard. New e2e checks: mic is off synchronously after stop (no wait for `onend`) and restarts immediately with a fresh recognizer.

File: web/chair/src/dictation.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  // are then dropped by the identity guard above. hard=true (submit) aborts +
+  // discards the engine buffer; graceful (mic tap) stop()s so nothing is pending.
+  const end = (hard: boolean): void => {
+    const self = rec;
+    if (!self) return;
+    rec = undefined;
+    handlers.onState(false);
+    try {
+      if (hard) self.abort();
+      else self.stop();
```
[the remark this answers is not in the document]
Comment (kdr) [reply web/chair/src/dictation.ts]: Fixed in 99d0a5b. This was the trade-off from round 2 (clearing `rec` synchronously on stop) — reconciled by decoupling the two concerns: `listening` (UI/toggle) flips off synchronously so the toggle stays responsive, while `rec` (the recognizer whose `onresult` we still accept) lingers until its `onend`, so the last interim→final promotion still lands. Hard `cancel()` clears `rec` immediately (submit still discards the buffer). New e2e check: a final result emitted after a graceful stop still updates the textarea, and a restart still works.

File: web/chair/src/views/composer.ts [hunk 6]
```diff
@@ -38,7 +42,51 @@ export function createComposer(handlers: {
...
+    }
+  });
+
+  let busy = false;
+  const busyPlaceholder = (): string => (busy ? "steer the running agent…" : "message the desk…");
+
   const submit = async (): Promise<void> => {
+    dictation.cancel();
```
[the remark this answers is not in the document]
Comment (kdr) [reply web/chair/src/views/composer.ts:100]: Fixed in 99d0a5b. Grouped with the auth-teardown report as one bucket — exit paths that must account for an active mic. `submit()` now locks the mic button (`mic.disabled = true`) for the duration of the async `onSend` and re-enables it in `finally`, mirroring the existing send-button lock, so a mic-start can't capture the outgoing text as `base` and restore it after the box clears. New e2e check (route-delayed send): the mic is disabled while the send is in flight and re-enabled after.

File: web/chair/src/main.ts [hunk 7]
```diff
@@ -87,6 +87,7 @@ async function boot(): Promise<void> {
         .then(refreshIfDisconnected)
         .catch((e) => (isAuthError(e) ? onAuthFailure() : transcript.notice("abort failed", "error")));
     },
+    onNotice: (text, level) => transcript.notice(text, level),
```
[the remark this answers is not in the document]
Comment (kdr) [reply web/chair/src/main.ts:91]: Fixed in 99d0a5b (same exit-path bucket as the mic-during-send report). The composer now exposes `cancelDictation()`, and `main.ts` calls it in `teardownSession()` — the single teardown primitive behind `onAuthFailure` — so the browser mic is aborted before the pairing gate renders. New e2e check (route-forced 401 → visibilitychange resync): the teardown shows the gate and the recognizer is aborted.

File: web/chair/src/dictation.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    } catch {
+      // some engines throw synchronously (e.g. InvalidStateError) with no onend
+      // to follow — roll the optimistic state back so the mic "on" UI can't
+      // stick (guarded so a racing newer session is left untouched)
+      if (current()) {
+        rec = undefined;
+        handlers.onState(false);
+      }
+      handlers.onError("could not start dictation — try again");
+    }
```
[the remark this answers is not in the document]
Comment (kdr) [reply web/chair/src/dictation.ts:138]: Fixed in 99d0a5b. The `start()` closure now tracks a per-session `sawError` flag set by `onerror`; the catch only emits the generic "could not start dictation" notice when `onerror` did not already surface a specific one, so one failed start = one notice. New e2e check: an `onerror`(not-allowed)+throw in a single start yields exactly one notice.

File: src/extension/chair.ts [hunk 9]
```diff
@@ -392,11 +429,29 @@ export function registerChair(pi: ExtensionAPI): ChairHandle {
...
   }
 
   async function startChair(opts: StartOptions = {}): Promise<void> {
+    // --serve: front the chair with HTTPS via `tailscale serve` so the paired
+    // phone loads a SECURE context and voice dictation works. serve proxies to
+    // loopback, so force a loopback bind; enable it before binding, and record
+    // the port so /chair off tears exactly this mapping down.
+    if (opts.serve) {
+      if (opts.bind === undefined) opts.bind = "127.0.0.1";
+      const port = opts.port ?? lastStartOpts.port ?? envPort(process.env.OVERCAST_CHAIR_PORT) ?? 7373;
```
[the remark this answers is not in the document]
Comment (kdr) [reply src/extension/chair.ts:455]: Fixed in b22eede. `--serve` now ALWAYS forces a loopback bind (`tailscale serve` proxies to loopback, so a tailnet bind would make the HTTPS QR point at a chair the proxy can't reach), and emits a note that it overrides a tailnet bind. Grouped with the other three serve findings as one ownership fix.

File: src/extension/chair.ts [hunk 10]
```diff
@@ -448,12 +503,20 @@ export function registerChair(pi: ExtensionAPI): ChairHandle {
...
+    const voiceLine = b.secure
+      ? "  🔒 HTTPS via Tailscale — voice dictation enabled on the phone"
+      : `  ⚠ HTTP — voice needs HTTPS: run \`${serveCommandHint(b.port)}\` then /chair qr, or /chair on --serve`;
     const lines = [
       "  ◉ CHAIR — scan to pair (token is in the QR only)",
       "",
-      ...qrLines(bridge.pairingUrl).map((l) => "  " + l),
+      ...qrLines(b.pairingUrl).map((l) => "  " + l),
       "",
-      `  ${bridge.url}   ·   /chair qr to hide`,
```
[the remark this answers is not in the document]
Comment (kdr) [reply src/extension/chair.ts:456]: Fixed in b22eede. `serveEnabledPort` is no longer set up front — it's adopted only AFTER `bindBridge` succeeds. If the bind fails, a serve mapping we just created is torn back down (so no orphaned HTTPS proxy to a dead port), and the previous-bridge restore path is unaffected.

File: src/extension/chair.ts [hunk 11]
```diff
@@ -392,11 +429,29 @@ export function registerChair(pi: ExtensionAPI): ChairHandle {
...
+    // loopback, so force a loopback bind; enable it before binding, and record
+    // the port so /chair off tears exactly this mapping down.
+    if (opts.serve) {
+      if (opts.bind === undefined) opts.bind = "127.0.0.1";
+      const port = opts.port ?? lastStartOpts.port ?? envPort(process.env.OVERCAST_CHAIR_PORT) ?? 7373;
+      emitResult(pi, "▶ chair: enabling HTTPS via `tailscale serve`…");
+      const r = await enableServe(port);
+      if (r.url) {
+        opts.publicUrl = r.url;
+        serveEnabledPort = port;
```
[the remark this answers is not in the document]
Comment (kdr) [reply src/extension/chair.ts]: Fixed in b22eede. `enableServe` now returns `created` — true only when it actually started the serve, false when it merely DETECTED an existing one. `serveEnabledPort` (and thus the `/chair off` teardown) is set only when `created` is true, so an operator's pre-existing serve is never torn down.

File: src/chair/serve.ts [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+  const detail = (res.stderr.trim() || res.stdout.trim()).split("\n").filter(Boolean)[0] || `exit ${res.code}`;
+  return {
+    error: `tailscale serve exposed no HTTPS URL (${detail}) — run \`${serveCommandHint(port)}\` manually, or enable HTTPS certificates + MagicDNS for your tailnet in the admin console`,
+  };
+}
+
+/** Best-effort teardown of the 443 serve mapping we brought up. */
+export async function disableServe(port: number): Promise<void> {
+  await runTailscale(["serve", "--https=443", "off"], 8000).catch(() => {});
+  void port;
```
[the remark this answers is not in the document]
Comment (kdr) [reply src/chair/serve.ts]: Fixed in b22eede. `tailscale serve --https=443 off` is coarse (whole 443), so `disableServe` now guards it with a new pure `serveIsSolelyPort` check: it only runs the off when EVERY serve handler proxies to the chair's loopback port. If any other mapping shares the 443 config it skips teardown and reports it (`… other tailscale serve mappings share the HTTPS config — left it up`).

File: web/chair/src/views/composer.ts [hunk 13]
```diff
@@ -38,16 +45,74 @@ export function createComposer(handlers: {
...
+  // Dictation: tap-to-toggle. While listening the textarea is the live
+  // transcript view (base text captured at start + finals + interim); a manual
+  // stop keeps whatever is displayed, submit hard-cancels so a late final
+  // flush can't repopulate the cleared box.
+  let base = "";
+  const dictation = createDictation({
+    onText: (finalText, interim) => {
+      const spoken = (finalText + interim).trim();
+      input.value = base ? (spoken ? `${base} ${spoken}` : base) : spoken;
+      autosize();
```
[the remark this answers is not in the document]
Comment (kdr) [reply web/chair/src/views/composer.ts:62]: Fixed in b22eede. A real `input` event on the textarea (typing/paste/IME) now cancels dictation, so a late interim/final can't rebuild the box from the frozen `base` and wipe the correction. This keys off the genuine `input` event — our own `onText` sets `.value` programmatically, which does NOT fire `input`, so normal transcript updates don't trip it. New e2e check: a manual edit mid-dictation stops the mic and a following onText leaves the edited text intact.

File: src/extension/chair.ts [hunk 14]
```diff
@@ -392,11 +429,35 @@ export function registerChair(pi: ExtensionAPI): ChairHandle {
     // store the RESOLVED bind + the ACTUAL bound port, so a reload rebinds to the
     // same concrete address the phone is paired to (an ephemeral 0 became a real
     // port; a partial rebind kept the prior bind)
-    lastStartOpts = { bind, port: b.port };
+    // Persist the explicit --url override for reloads; an auto-detected serve
+    // URL is re-resolved on each bind (cheap) so it stays correct if serve moves.
+    lastStartOpts = { bind, port: b.port, publicUrl: opts.publicUrl };
```
[the remark this answers is not in the document]
Comment (kdr) [reply src/extension/chair.ts]: Fixed in a0f5ed2. A partial `/chair on --port …` was dropping a previously-set `--url`. Root cause: `--serve` overloaded `opts.publicUrl` with a port-tied serve URL. Split them — the transient serve URL is now `opts.serveUrl` (NOT persisted; re-detected each bind), and the explicit `--url` stays in `opts.publicUrl`. bindBridge falls back to the last explicit `--url` both when resolving the origin AND when persisting `lastStartOpts`, so a partial rebind and a reload both keep it. New e2e-style unit test covers a partial rebind + reload preserving the origin (it caught a follow-on bug where the persist step overwrote the saved URL with undefined — fixed).

File: src/extension/chair.ts [hunk 15]
```diff
@@ -392,11 +429,35 @@ export function registerChair(pi: ExtensionAPI): ChairHandle {
...
+      const port = opts.port ?? lastStartOpts.port ?? envPort(process.env.OVERCAST_CHAIR_PORT) ?? 7373;
+      emitResult(pi, "▶ chair: enabling HTTPS via `tailscale serve`…");
+      const r = await enableServe(port);
+      if (r.url) {
+        opts.publicUrl = r.url;
+        if (r.created) serveCreatedPort = port; // ours to tear down — adopted only after bind succeeds
+      } else {
+        emitResult(pi, `▶ chair: ${r.error} — continuing over HTTP (voice needs HTTPS)`);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (kdr) [reply src/extension/chair.ts:471]: Fixed in a0f5ed2. `/chair on --url …` or `--serve` on an already-running bridge was short-circuiting to `showStatus()` (the bare-`/chair on` fast path) when bind/port were unchanged, so the new origin never applied. The bare check now also treats an origin change (`opts.publicUrl`/`opts.serveUrl` set) as a real change, forcing a rebind so HTTPS voice can be enabled without a full off/on. New unit test: a running chair adopts `--url` and its `displayUrl`/`secure` flip to the HTTPS origin.

File: src/extension/chair.ts [hunk 16]
```diff
@@ -392,17 +435,45 @@ export function registerChair(pi: ExtensionAPI): ChairHandle {
...
     const prevOpts = lastStartOpts; // the opts the running bridge is using (concrete)
     if (wasRunning) {
-      // explicit bind/port → rebind: stop (rotating the token unless pinned)
-      // and fall through to a fresh start; a bare `/chair on` just reports.
-      if (opts.bind === undefined && opts.port === undefined) {
+      // explicit bind/port/origin → rebind: stop (rotating the token unless
+      // pinned) and fall through to a fresh start; a bare `/chair on` just
+      // reports. `--url`/`--serve` are origin CHANGES, so they force a rebind
+      // too — otherwise HTTPS voice could never be enabled on a running chair.
+      if (opts.bind === undefined && opts.port === undefined && opts.publicUrl === undefined && opts.serveUrl === undefined) {
```
[the remark this answers is not in the document]
Comment (kdr) [reply src/extension/chair.ts:479]: Fixed in 8528a32. The loopback bind was being forced BEFORE `enableServe` finished, so a FAILED serve still tripped the rebind guard. Now `opts.bind` is set to loopback ONLY after serve succeeds — if serve fails, opts is left unchanged, so a bare `/chair on --serve` on a running (e.g. tailnet-bound) chair stays a no-op: no loopback yank, no token rotation. New unit test asserts a failed `--serve` leaves the running chair's bind/port/token untouched.

File: src/extension/chair.ts [hunk 17]
```diff
@@ -414,10 +485,14 @@ export function registerChair(pi: ExtensionAPI): ChairHandle {
     await stopping;
     const err = await bindBridge(opts);
     if (!err) {
+      if (serveCreatedPort !== undefined) serveEnabledPort = serveCreatedPort; // now there's a listener behind it
```
[the remark this answers is not in the document]
Comment (kdr) [reply src/extension/chair.ts:501]: Fixed in 8528a32. After a successful bind we now reconcile the serve mapping with the port actually bound: if we own a `tailscale serve` for a port the chair no longer listens on (a `--port` rebind moved off it), it's torn down (`serveIsSolelyPort`-guarded), and any serve created for the new bind is adopted. So a `--serve` then `/chair on --port …` no longer leaves an orphaned HTTPS mapping proxying to a dead port. New unit test (logging fake tailscale) asserts the old port's mapping is torn down on the rebind.
